src/main/sensors/battery.c

```diff
@@ -53,53 +61,106 @@ uint16_t batteryAdcToVoltage(uint16_t src)
     return ((uint32_t)src * batteryConfig->vbatscale * 33 + (0xFFF * 5)) / (0xFFF * 10);
 }
 
-#define BATTERY_SAMPLE_COUNT 8
-
-void updateBatteryVoltage(void)
+static void updateBatteryVoltage(void)
 {
-    static uint16_t vbatSamples[BATTERY_SAMPLE_COUNT];
-    static uint8_t currentSampleIndex = 0;
-    uint8_t index;
-    uint16_t vbatSampleTotal = 0;
+    uint16_t vbatSample;
+    uint16_t vbatFiltered;
 
     // store the battery voltage with some other recent battery voltage readings
-    vbatSamples[(currentSampleIndex++) % BATTERY_SAMPLE_COUNT] = vbatLatestADC = adcGetChannel(ADC_BATTERY);
-
-    // calculate vbat based on the average of recent readings
-    for (index = 0; index < BATTERY_SAMPLE_COUNT; index++) {
-        vbatSampleTotal += vbatSamples[index];
-    }
-    vbat = batteryAdcToVoltage(vbatSampleTotal / BATTERY_SAMPLE_COUNT);
+    vbatSample = vbatLatestADC = adcGetChannel(ADC_BATTERY);
+    vbatFiltered = (uint16_t)lowpassFixed(&lowpassFilter, vbatSample, VBATT_LPF_FREQ);
+    vbat = batteryAdcToVoltage(vbatFiltered);
 }
 
-batteryState_e calculateBatteryState(void)
+#define VBATTERY_STABLE_DELAY 40
+/* Batt Hysteresis of +/-100mV */
+#define VBATT_HYSTERESIS 1
+
+void updateBattery(void)
 {
-    if (vbat <= batteryCriticalVoltage) {
-        return BATTERY_CRITICAL;
+    updateBatteryVoltage();
+    
+    /* battery has just been connected*/
+    if(batteryState == BATTERY_NOTPRESENT && vbat > VBATT_DETECT)
+    {
+        /* Actual battery state is calculated below, this is really BATTERY_PRESENT */
+        batteryState = BATTERY_OK;
+        /* wait for VBatt to stabilise then we can calc number of cells
+        (using the filtered value takes a long time to ramp up) 
+        We only do this on the ground so don't care if we do block, not
+        worse than original code anyway*/
+        delay(VBATTERY_STABLE_DELAY);
```

Comment:
```
 if(batteryState == BATTERY_NOTPRESEN)
   if(vbat <= VBATT_DETECT) {
     batteryStabilisedTime = currentTime + VBATTERY_STABLE_DELAY;
   } else if(cmp32(currentTime, batteryStabilisedTime ) > 0) {
        batteryState = BATTERY_OK;
```
or something similar, at least variable naming should be better ... it even restarts the timeout when battery voltage drops below threshold momentarily ..